diesel_cli/src/migrations/mod.rs

```diff
@@ -273,55 +273,75 @@ fn redo_migrations<Conn, DB>(
     DB: Backend,
     Conn: MigrationHarness<DB> + Connection<Backend = DB> + 'static,
 {
-    let migrations_inner = |harness: &mut HarnessWithOutput<Conn, _>| -> Result<(), Box<dyn Error + Send + Sync + 'static>> {
-        let reverted_versions = if args.get_flag("REDO_ALL") {
-            harness.revert_all_migrations(migrations_dir.clone())?
-        } else {
-            let number = args.get_one::<u64>("REDO_NUMBER").unwrap();
-            (0..*number)
-                .filter_map(|_|{
-                    match harness.revert_last_migration(migrations_dir.clone()) {
-                        Ok(v) => {
-                            Some(Ok(v))
-                        }
-                        Err(e) if e.is::<MigrationError>() => {
-                            match e.downcast_ref::<MigrationError>() {
-                                // If n is larger then the actual number of migrations,
-                                // just stop reverting them
-                                Some(MigrationError::NoMigrationRun) => None,
-                                _ => Some(Err(e)),
+    let migrations = MigrationSource::<DB>::migrations(&migrations_dir)
+        .unwrap_or_else(handle_error)
+        .into_iter()
+        .map(|m| (m.name().version().as_owned(), m))
+        .collect::<HashMap<_, _>>();
+    let applied_migrations = conn.applied_migrations().unwrap_or_else(handle_error);
+    let versions_to_revert = if args.get_flag("REDO_ALL") {
+        &applied_migrations
+    } else {
+        let number = args.get_one::<u64>("REDO_NUMBER").unwrap();
+        let number = std::cmp::min(*number as usize, applied_migrations.len());
+        &applied_migrations[..number]
+    };
+    let should_use_not_use_transaction = versions_to_revert.iter().any(|v| {
+        migrations
+            .get(v)
+            .map(|m| !m.metadata().run_in_transaction())
+            .unwrap_or_default()
+    });
```

Comment:
A single migration having `metadata().run_in_transaction()` as false can cause all the migrations to not be run as a transaction. So if there's a failure in the middle of the redo operation, how are we restoring the DB to a consistent state?

Reply:
We cannot really restore the database into a consistent state in that case. If one of the migrations fails it aborts the process there. It's then up to the user to fix the problems.
Now that's not optimal, but on the other hand we cannot do much more without a single transaction.
Other variants that I considered:
* Requiring a `--allow-without-transaction` flag to allow running these redo operations. I've decided not to do that, as some backends (mysql) never use transactions for migrations (because you cannot run data definition commands in transactions there). Having such a flag would be kind of surprising then, as it would imply something different.
* Just run those migrations outside of a transaction that do require to be executed outside of a transaction. That has the problem that you need to split down the migration list in a much finer manner + it also might result in a database where only a partial set of migrations was reapplied.

Reply:
right that makes sense.